ffi/src/lib.rs

```diff
@@ -912,6 +920,20 @@ mod tests {
         Box::leak(boxed)
     }
 
+    #[no_mangle]
+    extern "C" fn allocate_str(kernel_str: KernelStringSlice) -> NullableCvoid {
+        let s: &str = unsafe { TryFromStringSlice::try_from_slice(&kernel_str).unwrap() };
+        let b = Box::new(s.to_string());
```

Comment:
Re all of this `*mut str` is a fat pointer, because slices need to hold their length. To keep the length bundled up I used a `String`. There's probably a more elegant way to do it, but since it's testing code I wasn't going to spend too long optimizing it.

Reply:
To expand a bit, you can totally get a `c_void` pointer from a `String`
<details>
<summary>Code</summary>

```rust
        let boxed_str = s.into_boxed_str();
        let raw_str_ptr: *mut str = Box::into_raw(boxed_str);
        let non_null = NonNull::new(raw_str_ptr).unwrap();
        let ptr = non_null.cast::<c_void>();
```

</details>

but once you have a `NonNull<c_void>` I don't think you can recover it, because:
```rust
  // ptr from above which is a NonNull<c_void>
   ptr.as_ptr() as *mut str
```

So you get into needing to basically recreate `KernelStringSlice`, since you need to store the pointer and the length separately. We could probably create a slice and leak a pointer to it, and then re-assemble, but just boxing and unboxing the string seems easier.

isn't legal: ``cannot cast thin pointer `*mut c_void` to wide pointer `*mut str` ``

Reply:
Wait... `str` is actually a slice, not just deref to slice?